File: plugin/src/websocket_server.ts
/**
 * WebSocket client for MCPAdobe UXP plugin
 * Handles real-time communication between the UXP plugin and MCP server
 */

import { executeAction } from './tools';

export interface WebSocketMessage {
  id: string;
  action: string;
  data?: any;
}

export interface WebSocketResponse {
  id: string;
  success: boolean;
  result?: any;
  error?: string;
}

export class WebSocketClient {
  private socket: WebSocket | null = null;
  private url: string = 'ws://localhost:8765';
  private statusCallback: (status: string, isConnected: boolean) => void;

  constructor(statusCallback: (status: string, isConnected: boolean) => void) {
    this.statusCallback = statusCallback;
  }

  setUrl(url: string): void {
    this.url = url;
  }

  async connect(): Promise<void> {
    try {
      if (this.socket) {
        this.disconnect();
      }

      this.updateStatus('Connecting...', false);
      
      // @ts-ignore: UXP WebSocket API
      this.socket = new WebSocket(this.url);
      console.log(`WebSocket connecting to ${this.url}`);

      this.socket.onopen = () => {
        this.updateStatus('Connected', true);
        console.log('WebSocket connected');
      };

      this.socket.onmessage = (event) => {
        console.log('Message received:', event.data);
        try {
          const message = JSON.parse(event.data) as WebSocketMessage;
          this.handleMessage(message);
        } catch (error) {
          console.error('Error parsing message:', error);
        }
      };

      this.socket.onclose = () => {
        this.updateStatus('Disconnected', false);
        console.log('WebSocket disconnected');
        this.socket = null;
      };

      this.socket.onerror = (error) => {
        console.error('WebSocket error:', error);
        this.updateStatus('Connection error', false);
        this.socket = null;
      };

    } catch (error) {
      console.error('Failed to connect:', error);
      this.updateStatus(`Connection failed: ${error instanceof Error ? error.message : String(error)}`, false);
      throw error;
    }
  }

  private updateStatus(status: string, isConnected: boolean): void {
    if (this.statusCallback) {
      this.statusCallback(status, isConnected);
    }
  }

  private async handleMessage(message: WebSocketMessage): Promise<void> {
    try {
      console.log(`Processing message: ${message.action}`, message);
      const result = await executeAction(message.action, message.data);
      this.sendResponse(message.id, { success: true, result });
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error);
      console.error(`Error processing message: ${errorMessage}`);
      this.sendResponse(message.id, { success: false, error: errorMessage });
    }
  }


  private sendResponse(messageId: string, response: Omit<WebSocketResponse, 'id'>): void {
    if (!this.socket || this.socket.readyState !== 1) { // 1 = OPEN
      console.error('Cannot send response: WebSocket not connected');
      return;
    }

    const fullResponse: WebSocketResponse = {
      id: messageId,
      ...response
    };
    
    try {
      this.socket.send(JSON.stringify(fullResponse));
    } catch (error) {
      console.error('Error sending response:', error);
    }
  }

  isConnected(): boolean {
    return this.socket !== null && this.socket.readyState === 1; // 1 = OPEN
  }

  disconnect(): void {
    if (this.socket) {
      try {
        this.socket.close();
      } catch (error) {
        console.error('Error closing WebSocket:', error);
      }
      this.socket = null;
      this.updateStatus('Disconnected', false);
    }
  }
}

File: server/websocket_server.py
"""
WebSocket server for MCPAdobe server
Handles real-time communication between the Python server and UXP plugin
"""

import asyncio
import json
import logging
import time
import uuid
import websockets
from typing import Dict, Any, Callable, Optional, List

logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(name)s - %(levelname)s - %(message)s')
logger = logging.getLogger(__name__)

class WebSocketServer:
    def __init__(self, host: str = "localhost", port: int = 8765):
        self.host = host
        self.port = port
        self.clients: List[websockets.WebSocketServerProtocol] = []
        self.action_handlers: Dict[str, Callable] = {}
        self._server = None
        self.message_dict: Dict[str, Any] = {}
        self.response_events: Dict[str, asyncio.Event] = {}

    async def _handle_message(self, websocket: websockets.WebSocketServerProtocol, message: str):
        """Handle incoming message from client"""
        try:
            data = json.loads(message)
            if not isinstance(data, dict):
                logger.error(f"Invalid message format: {message}")
                return
                
            logger.info(f"Received message: {data}")
            
            # Handle response from plugin
            if "id" in data and ("result" in data or "error" in data):
                logger.info(f"Received response: {data}")
                # Here you could route responses to waiting handlers
                self.message_dict[data["id"]] = data
                self.response_events[data["id"]].set()
                return

        except json.JSONDecodeError:
            logger.error(f"Invalid JSON: {message}")
        except Exception as e:
            logger.exception("Error handling message:")
    
    # async def _send_response(self, websocket: websockets.WebSocketServerProtocol, message_id: str, result: Any, success: bool = True):
    #     """Send response to client"""
    #     response = {
    #         "id": message_id,
    #         "success": success
    #     }
        
    #     if success:
    #         response["result"] = result
    #     else:
    #         response["error"] = result
            
    #     await websocket.send(json.dumps(response))
    #     logger.debug(f"Sent response: {response}")
    
    
    async def _client_handler(self, websocket: websockets.WebSocketServerProtocol, ):
        """Handle new client connection"""
        self.clients.append(websocket)
        client_id = str(id(websocket))
        logger.info(f"Client connected [id: {client_id}] (total: {len(self.clients)})")
        
        try:
            async for message in websocket:
                await self._handle_message(websocket, message)
        except websockets.exceptions.ConnectionClosed:
            logger.info(f"Client connection closed [id: {client_id}]")
        finally:
            self.clients.remove(websocket)
            logger.info(f"Client disconnected [id: {client_id}] (remaining: {len(self.clients)})")
    
    async def wait_for_response(self, message_id: str) -> Any:
        """Wait for a response from the plugin"""
        event = self.response_events[message_id] = asyncio.Event()
        timeout = 10

        try:
            await asyncio.wait_for(event.wait(), timeout=timeout)
            return self.message_dict[message_id]
        finally:
            self.response_events.pop(message_id, None)

    async def send_instruction(self, action: str = "", data: Any = None) -> str:
        """
        Send instruction to plugin
        
        Args:
            client: Specific client to send to, or None for first available
            action: Action to execute on plugin side
            data: Additional data for the action
            
        Returns:
            str: Message ID
        """
        if not self.clients:
            logger.warning("No clients connected, cannot send instruction")
            return ""
            
        target_client = self.clients[-1]
        
        message_id = str(uuid.uuid4())
        message = {
            "id": message_id,
            "action": action,
            "data": data or {}
        }
        
        json_message = json.dumps(message)
        try:
            await target_client.send(json_message)
            logger.info(f"Sent instruction: {action} with id: {message_id}")
        except Exception as e:
            logger.exception("Error sending instruction:")
            raise e
    
        try:
            response = await self.wait_for_response(message_id)
            return response
        except Exception as e:
            logger.exception("Error waiting for response:")
            raise e
        
    async def start(self):
        """Start the WebSocket server"""
        self._server = await websockets.serve(
            self._client_handler, self.host, self.port
        )
        logger.info(f"WebSocket server started at ws://{self.host}:{self.port}")
        
    async def stop(self):
        """Stop the WebSocket server"""
        if self._server:
            self._server.close()
            await self._server.wait_closed()
            self._server = None
            logger.info("WebSocket server stopped")

# Singleton instance
_server_instance = None

def get_server(host: str = "localhost", port: int = 8765) -> WebSocketServer:
    """Get the global WebSocket server instance"""
    global _server_instance
    if _server_instance is None:
        _server_instance = WebSocketServer(host, port)
    return _server_instance 